### ISSUE TITLE
TFS vstest missing attachments

### ISSUE THREAD

**Issue by jeremeguenther** [Microsoft/azure-pipelines-tasks#9611]
## Environment
- Server - TFS on-premises Version 16.131.28106.2
    
- Agent - Private: running on windows 10
Agent.OS | Windows_NT
Agent.OSVersion | 10.0.17134
Agent.Version | 2.136.1

## Issue Description

I run a large number of automated tests using TFS and VSTest every night.  During the tests I take screenshots whenever an error occurs.  In the last few weeks I have noticed that my screenshots are still happening and I can dig into the _work folder to find them, but they are no longer showing up in the Attachments area for a test run in TFS.

This problem sounds very similar to issue #7685 .  To the best of my knowledge only windows updates have occurred since the last time I remember seeing attachments in my failure results.

I am in the process of converting from CodedUI to Selenium, so about a third of my tests run under Selenium, while the rest run under CodedUI.  I first noticed the issue with the Selenium tests as that is where my focus currently is, however it seems to be an issue for all the test Releases.

Prior to this I ran this code, and it always attached the files correctly.
            var path = TestContext.ResultsDirectory;
            var filePath = Path.Combine(path, FileName);
            img.Save(filePath);

As part of my troubleshooting I have added this following line, which does correctly include the file in the test results .trx file now (they never were included in that file in the past), but they are still not showing up under the attachments area.
            TestContext.AddResultFile(filePath);

Is there any windows update or something else that could have triggered this?  Is there any known resolution to this issue?  Issue #7685 seemed to indicate it was resolved, however that issue was also referring to VSTS (I know the name has changed several times), and I am using TFS.

### Task logs

[Enable debug logging and please provide the zip file containing all the logs for a speedy resolution]

### Error logs

[Insert error from the logs here for a quick overview]

**Comment by prawalagarwal** [comment]
@jeremeguenther We will need some more information to analyze this. Can you attach your build logs with system.debug set as true in build variables?

Are you running tests on single agent or in Distributed mode?

**Comment by jeremeguenther** [comment]
This is a Release, not a build, but the build profile that the Release is running off of does have the Build Configuration set to debug.

I am running a single test agent.

I have attached the logs for the Release step that creates the attachment.
[5_Client Otm Stage CodedUI Tests.log](https://github.com/Microsoft/azure-pipelines-tasks/files/2885285/5_Client.Otm.Stage.CodedUI.Tests.log)

**Comment by jeremeguenther** [comment]
You will notice the IE Driver getting started many times.  I have found that necessary in my testing because IE crashes so frequently, the best way to avoid it crashing is to kill it and start a fresh instance frequently.

**Comment by prawalagarwal** [comment]
@jeremeguenther this is not a known issue and we will have to reproduce it on our end to figure out what is wrong here. Can you confirm that the attachment is present at the path specified in the trx file?

As a work around can you try turning on the Custom batch size in Advanced settings of the VsTest task and setting a high batch size (like 100). This triggers a different flow where results are published from a different code and might resolve the issue for you. Can you also provide us with a sample project where you can repro this?

**Comment by jeremeguenther** [comment]
Look at my original post for TFS version number.

When I run the tests locally, then the Test File links in the trx file work correctly.  However, when the test is run automatically by TFS then those links do not work, and looking at the xml, the links are not even close to correct.  To be clear, I have never had these links in the trx file before, so I am not sure if the bad links are a change or if they have always been this way.

The paths should be something like:
\\dcox-w10\c$\agent\_work\r3\a\TestResults\......
This is what is in the xml:
      <ResultFiles>
        <ResultFile path="DCOX-W10\MenuItemCheck_Fail_Business Profile.jpg" />
        <ResultFile path="DCOX-W10\MenuItemCheck_NavigateToUrl_0.jpg" />
      </ResultFiles>

It is several days old at this point and it looks like the test results have been cleaned out.  However, I have verified in the past that the .jpg files are actually present on disk; although the paths they got generated at did not seem to be consistent.  They were always somewhere inside the TestResults sub-folder, but under that there were a couple of different paths that would get generated.

I have set the batch size to 100, I will report back when I know what the impact is.

As for a sample project, I am not sure what you mean.  Just create a test project that generates screenshots, there is nothing special about it and it works locally.  The part that is failing is TFS, and I am not sure how to export the TFS configuration.

**Comment by jeremeguenther** [comment]
I just ran a test with batch size set to auto, and a second test with batch size set to 100.

Set to auto the test results were stored on the file system.  The test screenshot was stored in three different places, so three different copies of the same file.

Set to 100 the test results do not seem to have been stored on the file system.  At least no place that I can find them.  The drop folder, which is normally cleaned up after the test has been run still exists, almost as if the test failed to clean up after itself and save its results.

Doing a file compare between the two resulting trx files I notice these differences:

```
batch size 100:
  <TestSettings name="default" id="4d7d9e0c-696d-4805-8646-fcd612ed7c0c">
    <Execution>
      <TestTypeSpecific />
    </Execution>
    <Deployment runDeploymentRoot="SVC_TFS_DCOX-W10 2019-02-21 12_45_54" />
    <Properties />
  </TestSettings>

<TestMethod adapterTypeName="Microsoft.VisualStudio.TestTools.TestTypes.Unit.UnitTestAdapter"

batch size auto:
  <TestSettings name="default" id="ff2b1600-e382-4729-8918-bef2c0be92a0">
    <Deployment runDeploymentRoot="SVC_TFS_DCOX-W10_2019-02-21_12_10_06" />
  </TestSettings>

<TestMethod adapterTypeName="executor://mstestadapter/v2"
```

**Comment by prawalagarwal** [comment]
@jeremeguenther Can we schedule a call for this? 10AM to 8PM IST works for me.
You can set the system.debug variable in your release variables too and get the logs. They will have more details than the current ones.
Also, I could not reproduce this on my end, so we need to see what is different in your setup.

**Comment by jeremeguenther** [comment]
I tried adding System.Debug set to true as a Release variable, but it does not seem to have changed the output at all.

I emailed you my contact information, as well as an attachment with all the test results.  You will notice that when batch size is set to 100 none of the test result information gets written to the file system, not even the trx file.  You will also notice the test screenshot is saved three different times in three different paths when the batch size is set to auto.

I downloaded both trx files from TFS and included them with the zipped folders.  I was wrong before, the drop folders are always left behind after the tests have run, whether or not any test results are saved.

**Comment by prawalagarwal** [comment]
@jeremeguenther There is no way to add run level attachments from inside the test today. Also, missed to inform you in the call, you can access the individual tests from the Runs tab too and there no need to visit the summary page. You will see a tests tab on the top in the 'Runs' view which has all the tests listed in that run. Opening individual tests will show the attachments then.
Closing this issue now. Let me know if you have anymore concerns around this.

**Comment by jeremeguenther** [comment]
Ok, thanks for helping me with this.  It is odd that I used to get the attachments showing up as run level attachments.  But at least I do see them correctly showing up in the attachments area for each individual test.

**Comment by jeremeguenther** [comment]
@pagarwal-msft I upgraded my TFS to update 3.2.  And I have been doing some more testing.
So far, in my Selenium tests if I added the TestContext.AddResultFile(filePath); line then the screenshots show up in the test specific area, while if I remove that line then they are saved, but not listed anywhere in the test results.  This remains true regardless of the batch size being set to 100 or auto.

However, in my old Coded-UI tests that do not have that line in them the attachments have started showing up in the Run level area again.  Historical attachments have also shown up as well, so it must have been a display bug.

The odd part is that the Coded-UI stuff saves attachments to the Run level while Selenium does not.  This does not make sense to me.  There is definitely a difference somewhere in the Microsoft code between the two, i'm just not sure where.  The Run level attachments are definitely more convenient for me, however I can work with the test level attachments for now.

My guess is that Coded-UI actually has a bug in it allowing this since MS says it should not be possible.  I would like to request this as a feature for the future so I can get that functionality back.
